### ISSUE TITLE
IllegalStateException at dequeueOutputBuffer when transitioning from IMA ad to content

### ISSUE THREAD

**Issue by benliuxie** [google/ExoPlayer#4369]
### Issue description
IllegalStateException at dequeueOutputBuffer when transitioning from some IMA ad to content. And this only happens on some devices.

### Reproduction steps
Play DRM content with DemoApp. When IMA ad break completes content start to play 1 or 2 seconds with sound but video stays at last frame of AD. There is Internal runtime error IllegalStateException at android.media.MediaCodec.native_dequeueOutputBuffer(Native Method) (see logcat attached).

This seems to happen on some Ad only when transitioning from Ad to content. I reproduced the above steps with our own application, and it is reproducible on the Exoplayer Demo app with only the DRM flow update.

There is no problem if I only play the DRM content only.

### Link to test content
Its DRM content. I will send video url and adTag url by email, but we have some custom DRM calls for the playback to work.

### Version of ExoPlayer being used
2.7.1, 2.8.0 and 2.8.1

### Device(s) and version(s) of Android being used
Nexus 6 with Android 7.1.1
Galaxy Note 4 with Android 6.0.1
We cannot reproduce this on other devices such as Nexus 4, Nexus 5, Pixel and Pixel 2, and some other Galaxy phones and tablets.

### A full bug report captured from the device
[ExoPlayerDecodeError.txt](https://github.com/google/ExoPlayer/files/2091570/ExoPlayerDecodeError.txt)

**Comment by AquilesCanta** [comment]
Before we take a look into this, please provide a full bugreport as obtained from `adb bugreport`.

**Comment by benliuxie** [comment]
Thanks for responding so quickly. I could not run the command on Note 4 with Android 6 yesterday, please see the report from Nexus 6 with Android 7.1.1:
[bugreport-N6F27M-2018-06-12-10-05-28.zip](https://github.com/google/ExoPlayer/files/2094358/bugreport-N6F27M-2018-06-12-10-05-28.zip)

**Comment by AquilesCanta** [comment]
I think we'll also need to reproduce the issue. This will also allow us to reroute the problem easily if needed. 

Can you provide the required tokens/request headers to reproduce in the demo app?

Thanks.

**Comment by benliuxie** [comment]
Hi Santiago,



You can use the urls (same as I sent last time) to construct
content/ad mediaSource in PlayerActivity:

String videoUrl = "https://globaltvvideo.corusdigitaldev.com/globaltv/content/vod/Shaw_-_GlobalTV/950/859/1243315267743/1527533417_drm_24afd238-3ec3-4132-8d09-b096b88ad983.ism/.mpd";String adTagUrl = "https://pubads.g.doubleclick.net/gampad/ads?sz=320x240&output=xml_vmap1&impl=s&gdfp_req=1&env=vp&unviewed_position_start=1&ad_rule=1&idtype=adid&correlator=1528732382&iu=/6872/corus.globaltvdai.com&vid=1243315267743&cmsid=2468523&ppid=15283129164091148018914&cust_params=show%3DBig+Brother+Canada%26description_url%3Dhttp%3A%2F%2Fwww.globaltv.com”;
And use the following request(see hard-coded key request url and payload/mediaToken as well as header) in the Widevine DrmCallback:

String mediaoToken = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJkdWlkIjoiNzkyOGFiYjYtYjcyYy00MWVhLThiNGUtYmE1N2NhNGE3NmMyIiwicmVzb3VyY2VfaWQiOiIxMjQzMzE1MjY3NzQzIiwiZXhwIjoxNTI4ODMyMTc0LCJzdWlkIjoiMWY2MjFmODQtYTY5ZS00YWU4LWFhZTUtODUxYmViYzg2NjIyIn0.WnJzFHH2Sb_LleIZcijMT9cfBMGtvDs8Z3Lgwkb0XCk";@Overridepublic byte[] executeKeyRequest(UUID uuid, ExoMediaDrm.KeyRequest request) throws Exception {    String url = "https://global.corusappservices.com/authorization/widevine/getresourcekey";    String requestStr = getWidewineRequestBody(request.getData(), mediaoToken);    byte[] result =  executePost(dataSourceFactory, url, requestStr.getBytes(), null);    return result;}private String getWidewineRequestBody(byte[] requestData, String authzToken){    String result =null;    try {        JSONObject jsonObject = new JSONObject();        jsonObject.put("authorization_token", authzToken);        JSONArray array = new JSONArray();        for(int i= 0; i< requestData.length; i++){            int x = 0 ;            x = ( requestData[i] & 0xFF) ;            array.put(x);        }        jsonObject.put("license_request_data", array);        result = jsonObject.toString();    }catch (Exception e){        e.printStackTrace();    }    return result;}private static byte[] executePost(HttpDataSource.Factory dataSourceFactory, String url,                                  byte[] data, Map<String, String> requestProperties) throws IOException {    HttpDataSource dataSource = dataSourceFactory.createDataSource();    dataSource.setRequestProperty("Content-Type", "application/json”);Note the Token expires in 2 hours, its hard for us to extend this so I need to send new one if you need it.Thanks,Ben

**Comment by benliuxie** [comment]
I have sent the information by email. Its very easy to reproduce on the Global ad (Shade of Blue) and History Ad (Alone), almost 100% reproducible when transitioning to content from these ads. I thought it is related to the ads but I tried to limit the bit rate, or content-type this issue still happens, and error happens at beginning of content playback: with last frame of ad then black screen.

Thanks,
Ben

**Comment by benliuxie** [comment]
Pixel 1 randomly reboot when transitioning from these ads to content too. Sometimes playing a few video in couple of hours without problem but sometimes device reboots more frequently. I don't know if its related because there is different exception in the logs. I have attached bugreports for Pixel 1 and hope it helps.
[bugreport-sailfish-OPM1.171019.021-2018-06-12-15-57-38.zip](https://github.com/google/ExoPlayer/files/2095669/bugreport-sailfish-OPM1.171019.021-2018-06-12-15-57-38.zip)

**Comment by benliuxie** [comment]
Limit the content bitrate and choose EXTENSION_RENDERER_MODE_ON seem to have fixed this issue. This is still under QA test so let us know if you have other suggestions. 

Thanks,
Ben

**Comment by andrewlewis** [comment]
Looking at the logcat, it's surprising that the exception is thrown about ten seconds after switching codecs. Maybe the stream is invalid ten seconds after the transition (there isn't much information to go on, so this is really a guess!).

Are you sure turning on extension renderers makes a difference? The formats mentioned in the logcat are VP8 + Vorbis and H.264 + AAC, so the only streams that might be handled by extension renderers are audio.

**Comment by benliuxie** [comment]
I have tried many settings, initially focusing on Ad/IMA since this happens only for some ads transitioning to content. I started looking into the content related settings yesterday. Lowering the max bit rate did make a difference. I found it is less easily to reproduce with some render setting change before trying the bit rate so this is a guess too. I turn it on just in case the core one failed so it has the option to uses the extension.
